core/src/main/java/hudson/FilePath.java

```diff
@@ -1400,17 +1403,25 @@ public String invoke(File dir, VirtualChannel channel) throws IOException {
      * @return
      *      The new FilePath pointing to the temporary directory
      * @since 1.311
-     * @see File#createTempFile(String, String)
+     * @see Files#createTempDirectory(Path, String, FileAttribute[])
      */
     public FilePath createTempDir(final String prefix, final String suffix) throws IOException, InterruptedException {
         try {
+            String[] s;
+            if (StringUtils.isBlank(suffix)) {
+                s = new String[]{prefix, "tmp"}; // see File.createTempFile - tmp is used if suffix is null
+            } else {
+                s = new String[]{prefix, suffix};
+            }
+            String name = StringUtils.join(s, ".");
             return new FilePath(this,act(new SecureFileCallable<String>() {
                 private static final long serialVersionUID = 1L;
                 public String invoke(File dir, VirtualChannel channel) throws IOException {
-                    File f = File.createTempFile(prefix, suffix, dir);
-                    f.delete();
-                    f.mkdir();
-                    return f.getName();
+                    Path tempPath = Files.createTempDirectory(Util.fileToPath(dir), name,  new FileAttribute<?>[] {});
```

Comment:
Found a potentially relevant issue while investigating [JENKINS-48407](https://issues.jenkins-ci.org/browse/JENKINS-48407). The old io code would create directories with permissions `0777 & (~umask)`, whereas the nio code will create directories with permissions `0700 & (~umask)`. It's easy to retain the old behavior by passing `PosixFileAttributes.asFileAttributes(EnumSet.allOf(PosixFilePermission.class))` as the file attribute parameter to `Files#createTemporaryDirectory` after checking `Functions#isWindows`. CC @oleg-nenashev do you think it is important to maintain the old permissions here?

Reply:
Better safe than sorry. I would vote for retaining the permissions in this case for sure